### ISSUE TITLE
Worker process crashes after bazel clean

### ISSUE THREAD

**Issue by davido** [bazelbuild/bazel#2769]
/CC @hanwen @damienmg 

I'm using Bazel 0.4.5. To reproduce, use Gitiles project: [1].

The sequence `bazel build gitiles; bazel clean; bazel build gitiles` consistently leads to worker process crash:

```
$ bazel build gitiles
.
INFO: Found 1 target...
Target //:gitiles up-to-date:
  bazel-bin/gitiles.war
INFO: Elapsed time: 8.867s, Critical Path: 5.87s

$ bazel clean --async 
INFO: Starting clean.
INFO: Output tree moved to /home/davido/.cache/bazel/_bazel_davido/9fdffb7f2a398de2e1a63eaee2ea7c50/execroot/gitiles_tmp_18801 for deletion

$ bazel build gitiles
INFO: Found 1 target...
ERROR: /home/davido/projects/gitiles/blame-cache/BUILD:8:1: Java compilation in rule '//blame-cache:lib' failed: Worker process sent response with exit code: 1.
java.io.IOException: Cannot write dependencies to bazel-out/local-fastbuild/bin/blame-cache/liblib.jdeps
	at com.google.devtools.build.buildjar.javac.plugins.dependency.DependencyModule.emitDependencyInformation(DependencyModule.java:137)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.run(SimpleJavaLibraryBuilder.java:188)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.processRequest(BazelJavaBuilder.java:90)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.runPersistentWorker(BazelJavaBuilder.java:67)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.main(BazelJavaBuilder.java:44)
Caused by: java.io.FileNotFoundException: bazel-out/local-fastbuild/bin/blame-cache/liblib.jdeps (No such file or directory)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:101)
	at com.google.devtools.build.buildjar.javac.plugins.dependency.DependencyModule.emitDependencyInformation(DependencyModule.java:133)
	... 4 more
Target //:gitiles failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 2.914s, Critical Path: 1.06s
```

Workaround: killing the worker process with `bazel shutdown` fixes the problem:

```
$ bazel shutdown

$ bazel build gitiles
.
INFO: Found 1 target...
Target //:gitiles up-to-date:
  bazel-bin/gitiles.war
INFO: Elapsed time: 8.641s, Critical Path: 5.41s
```

* [1] git clone https://gerrit.googlesource.com/gitiles

**Comment by mhlopko** [comment]
We had the same problem on our ci, and it got away after bazel clean --expunge. Can you repro on a fresh clone?

**Comment by hanwen** [comment]
we have a workaround already, but if we need to restart things and/or run "bazel clean", then something is wrong with bazel.

**Comment by mhlopko** [comment]
Definitely true, I was hoping I would get a repro. I'm still investigating the cause.

**Comment by davido** [comment]
Confirmed, that using `--expunge` option fixed it. I also created a simple repdoducer repository with one single source file only: https://github.com/davido/bazel_printy:

```
  $ git clone https://github.com/davido/bazel_printy bazel_clean_breakage && cd bazel_clean_breakage
  $ bazel build printy
........
INFO: Found 1 target...
Target //:printy up-to-date:
  bazel-bin/printy.jar
  bazel-bin/printy
INFO: Elapsed time: 5.501s, Critical Path: 1.72s

  $ bazel clean --async 
INFO: Starting clean.
INFO: Output tree moved to /home/davido/.cache/bazel/_bazel_davido/b556e72e29ea9d91a4088eef654d01a3/execroot/bazel_clean_breakage_tmp_25352 for deletion

  $ bazel build printy
INFO: Found 1 target...
ERROR: /home/davido/projects/bazel_clean_breakage/BUILD:14:1: Java compilation in rule '//:printy' failed: Worker process sent response with exit code: 1.
java.io.FileNotFoundException: bazel-out/local-fastbuild/bin/printy.jar (No such file or directory)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:101)
	at com.google.devtools.build.buildjar.jarhelper.JarCreator.execute(JarCreator.java:176)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.buildJar(SimpleJavaLibraryBuilder.java:230)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.run(SimpleJavaLibraryBuilder.java:180)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.processRequest(BazelJavaBuilder.java:90)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.runPersistentWorker(BazelJavaBuilder.java:67)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.main(BazelJavaBuilder.java:44)
Target //:printy failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 2.079s, Critical Path: 0.34s
```

Now, when I tried to use that option `--expunge-async`, it crashes too:

```
$ bazel clean --expunge_async 
INFO: Starting clean.
INFO: Output base moved to /home/davido/.cache/bazel/_bazel_davido/b556e72e29ea9d91a4088eef654d01a3_tmp_27565 for deletion
java.util.logging.ErrorManager: 4
java.io.FileNotFoundException: /home/davido/.cache/bazel/_bazel_davido/b556e72e29ea9d91a4088eef654d01a3/java.log (No such file or directory)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:133)
	at java.util.logging.FileHandler.open(FileHandler.java:210)
	at java.util.logging.FileHandler.rotate(FileHandler.java:661)
	at java.util.logging.FileHandler.access$100(FileHandler.java:150)
	at java.util.logging.FileHandler$1.run(FileHandler.java:693)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.util.logging.FileHandler.publish(FileHandler.java:690)
	at java.util.logging.Logger.log(Logger.java:738)
	at java.util.logging.Logger.doLog(Logger.java:765)
	at java.util.logging.Logger.log(Logger.java:788)
	at java.util.logging.Logger.info(Logger.java:1490)
	at com.google.devtools.build.lib.runtime.commands.CleanCommand.asyncClean(CleanCommand.java:191)
	at com.google.devtools.build.lib.runtime.commands.CleanCommand.actuallyClean(CleanCommand.java:221)
	at com.google.devtools.build.lib.runtime.commands.CleanCommand.exec(CleanCommand.java:159)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:479)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:294)
	at com.google.devtools.build.lib.runtime.CommandExecutor.exec(CommandExecutor.java:47)
	at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:838)
	at com.google.devtools.build.lib.server.GrpcServerImpl.access$2100(GrpcServerImpl.java:101)
	at com.google.devtools.build.lib.server.GrpcServerImpl$4$1.run(GrpcServerImpl.java:898)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

With only expunge it worked in the end:

```
  $ bazel clean --expunge
........
INFO: Starting clean (this may take a while). Consider using --expunge_async if the clean takes more than several minutes.

  $ bazel build printy
INFO: Found 1 target...
Target //:printy up-to-date:
  bazel-bin/printy.jar
  bazel-bin/printy
INFO: Elapsed time: 5.678s, Critical Path: 1.20s
```

**Comment by mhlopko** [comment]
Perfect, I can repro now! Thanks David!

**Comment by philwo** [comment]
tl;dr - we have to kill workers on "bazel clean" to fix this. I don't think there's a way we can keep them running without extending the protocol with a "cwd" field for each WorkRequest and fixing all workers to explicitly chdir into that directory before opening any files. I'll send a fix for the clean command.

Workaround: Until this is fixed, use either "bazel clean" without any flags (though that might still cause failures, see below), or to be on the safe side, just use "bazel clean --expunge".

What's happening:

The working directory of persistent worker processes is the execution root (cf. "bazel info execution_root"), unless the flag --worker_sandboxing is used, in which case it is "$output_base/bazel-workers/worker-$id-$mnemonic".

The reason why you see these failures is that the working directory of the running persistent workers is deleted under their feet by the "clean" command, so the next time they try to open a file, it'll fail.

Let's look at each case individually:

bazel clean: Bazel will delete the execution_root directory. The Bazel server keeps running. The next time "bazel build" runs, a new execution_root directory is created in the same path as the old one. In theory this should invalidate the current working directory of the worker process (because while the path is the same, the inode is different), but JavaBuilder doesn't seem to mind and continues working fine. I guess other workers might fail in this case, too.

bazel clean --async: Bazel will move the execroot and thus the current working directory. You can actually observe this using the "pwdx" tool on the running JavaBuilder processes:

```
philwo@philwo:~/bazel-test$ bazel clean --async
INFO: Reading 'startup' options from /usr/local/google/home/philwo/.bazelrc: --watchfs
INFO: Starting clean.
INFO: Output tree moved to /usr/local/google/home/philwo/.cache/bazel/_bazel_philwo/11ae71351ba9bcc6fe0988005a8d1396/execroot/bazel-test_tmp_30646 for deletion
philwo@philwo:~/bazel-test$ pwdx 1876
1876: /usr/local/google/home/philwo/.cache/bazel/_bazel_philwo/11ae71351ba9bcc6fe0988005a8d1396/execroot/bazel-test_tmp_30646 (deleted)
```

The next time the worker tries to open a file, it will find itself in a no longer existing working directory and will crash.

bazel --expunge / --expunge_async: Work, because the Bazel server quits after the command, thus all workers quit, too. I don't know why --expunge_async doesn't work - maybe it's generally buggy? But that is a different bug than the worker process crash here (the shown stack trace comes from inside Bazel itself).

**Comment by hanwen** [comment]
oh, I had completely missed that there was a "bazel clean" in the middle. It's logical that cleaning things (asynchronously!) from under the worker's feet will lead to weirdness.

David, why are you running clean in the first place?

**Comment by davido** [comment]
@philwo Thank you for the details explanation. I haven only one question: What changed, why it is starting to fail now, or was `bazel clean` always broken?

> David, why are you running clean in the first place?

This is an absolute exception here, I was doing this while updating bazlets. In Gitiles or Gerrit we are using bazlets, with `git_repository`. Now, while working with bazlets code, I found, that only changing the bazlets source wouldn't re-fetch the bazlets. That why I used clean and discover with issue.

**Comment by hanwen** [comment]
"changing the bazlets source wouldn't re-fetch the bazlets" 

what does that mean? Are you writing self-modifiying code?

**Comment by davido** [comment]
> "changing the bazlets source wouldn't re-fetch the bazlets"

Two repositories:

* gitiles
* bazlets

Consider these CLs: [1], [2]. The former exposed `pkg_war` bazlet and the latter was using it. While working on it, I routed the bazlets to my local directory, as the commit wasn't yet pushed to bazlet to be consumed from:

```
  load_bazlets(
    commit = "a8a04c9ef2f131816c1e458613bee93b461ba652",
    local_path = "/home/davido/projects/bazlets",
)
```

After I have built it once, and `pkg_war` was fetched from the local bazlet repository located in `/home/davido/projects/bazlets`, i changed the `tools/pkg_war.bzl`, and re-built gitiles war in gitiles directory:

```
  $ cd gitiles
  $ bazel build gitiles
```

To make sure, that pkg_war bazlets is get re-fetched again, I issued `bazel clean && bagel build gitiles`.

* [1] https://gerrit-review.googlesource.com/101150
* [2] https://gerrit-review.googlesource.com/101071

**Comment by hanwen** [comment]
but was that necessary? Did it not work without make clean? The bazlets have a commit SHA1, so it should pick up changes correctly.

**Comment by davido** [comment]
I haven't committed anything in bazlets repository and haven't adapted any thing in bazlets. Note, that when `local_path` is passed to `load_bazlets` macro, then `git_repository` without SHA1 is used: [1].

```
  native.local_repository(
       name = NAME,
       path = local_path,
   )
```

* [1] https://gerrit.googlesource.com/gitiles/+/f9bb416b3d35ee00bda17360e4d74ec8f6350d58/tools/bazlets.bzl#14
